x/wasm/keeper/query_plugins.go

```diff
@@ -499,7 +500,23 @@ func WasmQuerier(k wasmQueryKeeper) func(ctx sdk.Context, request *wasmvmtypes.W
 			}
 			return json.Marshal(res)
 		}
-		return nil, wasmvmtypes.UnsupportedRequest{Kind: "unknown WasmQuery variant"}
+		return nil, wasmvmtypes.UnsupportedRequest{Kind: "unknown wasm query variant"}
+	}
+}
+
+// LimitedDepthWasmQuerier decorator to WasmQuerier to limit the depth of recursive calls
+func LimitedDepthWasmQuerier(maxDepth uint32, next func(ctx sdk.Context, request *wasmvmtypes.WasmQuery) ([]byte, error)) func(ctx sdk.Context, request *wasmvmtypes.WasmQuery) ([]byte, error) {
+	return func(ctx sdk.Context, request *wasmvmtypes.WasmQuery) ([]byte, error) {
+		counter, ok := types.QueryDepthCounter(ctx)
```

Comment:
ah, storing this in context is nice to handle this over multiple execution paths.

but what are the requirements / expected behaviour? can you document that first?

e.g. If there is a limit of 10 depths, and I dispatch 3 wasm messages and each make 4 queries, I do believe this will trigger the issue, even if not nested.

But... I am not really sure where these contexts will get unwrapped, so maybe this only will trigger when Contract 1 -> contract 2 -> ... Contract 11  and not multiple messages using the querier one after another.